server/hotrod/src/main/java/org/infinispan/server/hotrod/HotRodExceptionHandler.java

```diff
@@ -0,0 +1,44 @@
+package org.infinispan.server.hotrod;
+
+import io.netty.buffer.Unpooled;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.handler.codec.DecoderException;
+import io.netty.util.CharsetUtil;
+import org.infinispan.commons.logging.LogFactory;
+import org.infinispan.server.core.logging.JavaLog;
+import scala.Tuple2;
+
+import javax.security.sasl.SaslException;
+import java.io.IOException;
+import java.nio.channels.ClosedChannelException;
+import java.security.PrivilegedActionException;
+
+/**
+ * // TODO: Document this
+ *
+ * @author wburns
+ * @since 9.0
+ */
+public class HotRodExceptionHandler extends ChannelInboundHandlerAdapter {
+   private final static JavaLog log = LogFactory.getLog(HotRodExceptionHandler.class, JavaLog.class);
+
+   @Override
+   public void exceptionCaught(ChannelHandlerContext ctx, Throwable t) throws Exception {
+      Channel ch = ctx.channel();
+      NewHotRodDecoder decoder = ctx.pipeline().get(NewHotRodDecoder.class);
+      CacheDecodeContext decodeCtx = decoder.decodeCtx;
+
+      log.debug("Exception caught", t);
+      if (t instanceof DecoderException) {
+         t = t.getCause();
+      }
+      if (t instanceof HotRodException) {
+         // HotRodException is already translated to response
+         ch.writeAndFlush(((HotRodException) t).response());
+      } else {
+         ch.writeAndFlush(decodeCtx.createExceptionResponse(t));
+      }
```

Comment:
According to Netty's best practices, when doing a write and flush and not interested in the future, it's better to use ```ch.writeAndFlush(whatever, ch.voidPromise())``` to reduce allocation rates since it reuses the same promise. I wonder if we could adopt it a rule on all writes without breaking anything